packages/cdk8s/lib/dependency.ts

```diff
@@ -0,0 +1,195 @@
+ * childless children of the root node of the graph.
+ *
+ * The graph does not include cross-scope dependencies. That is, if a child on the current scope depends on a node
+ * from a different scope, that relationship is not represented in this graph.
+ *
+ */
+export class DependencyGraph {
+
+  private readonly _fosterParent: DependencyVertex;
+
+  constructor(node: Node) {
+
+    this._fosterParent = new DependencyVertex(null);
+
+    const nodes: Record<string, DependencyVertex> = {};
+
+    function putVertex(construct: IConstruct) {
+      nodes[Node.of(construct).uniqueId] = new DependencyVertex(construct);
+    }
+
+    function getVertex(construct: IConstruct): DependencyVertex {
+      return nodes[Node.of(construct).uniqueId];
+    }
+
+    // create all vertices of the graph.
+    for (const n of node.findAll()) {
+      putVertex(n);
+    }
+
+    // create all the edges of the graph.
+    for (const dep of node.dependencies) {
+
+      if (!node.tryFindChild(Node.of(dep.target).id)) {
+        // dont cross scope boundaries.
+        // since charts only renders its own children, this is ok and
+        // has the benefit of simplifying the graph. we should reconsider this behavior when moving
+        // to a more general purpose use-case.
+        continue;
+      }
+
+      const sourceDepNode = getVertex(dep.source);
+      const targetDepNode = getVertex(dep.target);
+
+      sourceDepNode.addChild(targetDepNode);
+
+    }
+
+    // create the root.
+    for (const n of Object.values(nodes)) {
+      if (n.parents.size === 0) {
+        // orphans are dependency roots. lets adopt them!
+        this._fosterParent.addChild(n);
+      }
+    }
+
+  }
+
+  /**
+   * Returns the root of the graph.
+   *
+   * Note that this vertex will always have `null` as its `.value` since it is an artifical root
+   * that binds all the connected spaces of the graph.
+   */
+  public get root(): DependencyVertex {
+    return this._fosterParent;
+  }
+
+  /**
+   * @see Vertex.topology()
+   */
+  public topology(): IConstruct[] {
+    return this._fosterParent.topology();
+  }
+}
+
+/**
+ * Represents a vertex in the graph.
+ *
+ * The value of each vertex is an `IConstruct` that is accessible via the `.value` getter.
+ */
+export class DependencyVertex {
+
+  private readonly _value: IConstruct | null;
+  private readonly _children: Set<DependencyVertex> = new Set<DependencyVertex>();
+  private readonly _parents: Set<DependencyVertex> = new Set<DependencyVertex>();
+
+  constructor(value: IConstruct | null) {
+    this._value = value;
+  }
+
+  /**
+   * Returns the IConstruct this graph vertex represents.
+   *
+   * `null` in case this is the root of the graph.
+   */
+  public get value(): IConstruct | null {
+    return this._value;
+  }
+
+  /**
+   * Returns the children of the vertex (i.e dependencies)
+   */
+  public get children(): Set<DependencyVertex> {
```

Comment:
Terminology: perhaps use “inbound” (dependents) and “outbound” (dependencies)

Reply:
Yeah lets go with that 👍